File: RubyMotionIOS/app/gate.rb
#
# Copyright (c) 2015 Harry Cheung
#

class Gate < Point

  START_FINISH = 'START_FINISH'
  START = 'START'
  FINISH = 'FINISH'
  SPLIT = 'SPLIT'

  LINE_WIDTH = 80
  BEARING_RANGE = 90

  attr_reader :gate_type, :split_number

  def initialize(gate_type, split_number, latitude, longitude, bearing)
    super(latitude, longitude, bearing: bearing)

    @gate_type = gate_type
    @split_number = split_number
    left_bearing = bearing - 90 < 0 ? bearing + 270 : bearing - 90
    right_bearing = bearing + 90 > 360 ? bearing - 270 : bearing + 90
    @left_point = destination(left_bearing, LINE_WIDTH / 2)
    @right_point = destination(right_bearing, LINE_WIDTH / 2)
  end

  def cross(start, destination)
    path_bearing = start.bearing_to(destination)
    cross = nil
    if path_bearing > bearing - BEARING_RANGE and
        path_bearing < bearing + BEARING_RANGE
      cross = Point.intersect_simple(@left_point, @right_point, start, destination)
    end
    unless cross.nil?
      distance = start.distance_to(cross)
      acceleration = (destination.speed - start.speed) / (destination.timestamp - start.timestamp)
      time = Physics.time(distance, start.speed, acceleration)
      cross.generated = true
      cross.speed = start.speed + acceleration * time
      cross.bearing = start.bearing_to(destination)
      cross.timestamp = start.timestamp + time
      cross.lap_distance = start.lap_distance + distance
      cross.lap_time = start.lap_time + time
      cross.split_time = start.split_time + time
    end
    cross
  end

end

File: SwiftApp/SwiftApp/Gate.swift
//
//  Copyright (c) 2015 Harry Cheung
//

import UIKit

enum GateType: String, Printable {
  case SPLIT = "SPLIT", START = "START", FINISH = "FINISH", START_FINISH = "START_FINISH"
  
  var description : String {
    switch self {
    case .SPLIT: return "SPLIT"
    case .START: return "START"
    case .FINISH: return "FINISH"
    case .START_FINISH: return "START_FINISH"
    }
  }
}


final class Gate {
  
  let LINE_WIDTH:    Double = 30
  let BEARING_RANGE: Double = 5
  var location: Point
  let type: GateType
  let splitNumber: Int
  var leftPoint, rightPoint: Point?
  
  init(type: GateType, splitNumber: Int, latitude: Double, longitude: Double, bearing: Double) {
    self.type = type
    self.splitNumber = splitNumber
    self.location = Point(latitude: latitude, longitude: longitude, inRadians: false)
    let leftBearing  = bearing - 90 < 0 ? bearing + 270 : bearing - 90
    let rightBearing = bearing + 90 > 360 ? bearing - 270 : bearing + 90
    self.leftPoint  = location.destination(leftBearing, distance: LINE_WIDTH / 2)
    self.rightPoint = location.destination(rightBearing, distance: LINE_WIDTH / 2)
    self.location.bearing = bearing
  }
  
  func crossed(#start: Point, destination: Point, inout cross: Point) -> Bool {
    let pathBearing = start.bearingTo(destination)
    if pathBearing > (location.bearing - BEARING_RANGE) &&
      pathBearing < (location.bearing + BEARING_RANGE) {
      if Point.intersectSimple(p: leftPoint!, p2: rightPoint!, q: start, q2: destination, intersection: &cross) {
        let distance     = start.distanceTo(cross)
        let timeSince    = destination.timestamp - start.timestamp
        let acceleration = (destination.speed - start.speed) / timeSince
        let timeCross    = Physics.time(distance: distance, velocity: start.speed, acceleration: acceleration)
        cross.generated   = true
        cross.speed       = start.speed + acceleration * timeCross
        cross.bearing     = start.bearingTo(destination)
        cross.timestamp   = start.timestamp + timeCross
        cross.lapDistance = start.lapDistance + distance
        cross.lapTime     = start.lapTime + timeCross
        cross.splitTime   = start.splitTime + timeCross
        return true
      }
    }
    return false
  }
}
